src/SAML2/XML/ds/X509Certificate.php

```diff
@@ -60,24 +51,35 @@ public function getCertificate(): string
      * @param string $certificate
      * @return void
      */
-    public function setCertificate(string $certificate): void
+    private function setCertificate(string $certificate): void
     {
         $this->certificate = $certificate;
     }
 
 
+    /**
+     * Convert XML into a X509Certificate
+     *
+     * @param \DOMElement $xml The XML element we should load
+     * @return self
+     */
+    public static function fromXML(DOMElement $xml): object
+    {
+        return new self($xml->textContent);
+    }
+
+
     /**
      * Convert this X509Certificate element to XML.
      *
-     * @param \DOMElement $parent The element we should append this X509Certificate element to.
+     * @param \DOMElement|null $parent The element we should append this X509Certificate element to.
      * @return \DOMElement
      *
      * @throws \InvalidArgumentException if assertions are false
      */
-    public function toXML(DOMElement $parent): DOMElement
+    public function toXML(DOMElement $parent = null): DOMElement
     {
-        Assert::notEmpty($this->certificate);
-
-        return Utils::addString($parent, XMLSecurityDSig::XMLDSIGNS, 'ds:X509Certificate', $this->certificate);
+//        $e = $this->instantiateParentElement($parent);
+        return Utils::addString($parent, self::NS, 'ds:X509Certificate', $this->certificate);
```

Comment:
This is one outstanding issue that occurs twice in this PR..
Our interface requires `toXML` to accept null. so I had to change it..
Passing `$e` to `Utils::addString` didn't work because it leads to `<ds:X509Certificate><ds:X509Certificate>somecert</ds:X509Certificate></ds:X509Certificate>`..

Reply:
Maybe time for `SAML2\Utils::appendString()`? Then both `SAML2\Utils::addString()` and `SAML2\Utils::addStrings()` could be reimplemented to use that one, and here we could use it too.